hail/python/hailtop/batch/batch.py

```diff
@@ -149,6 +151,58 @@ def __init__(self,
 
         self._cancel_after_n_failures = cancel_after_n_failures
 
+        self._python_functions: Dict[int, Tuple[Callable, Optional[str]]] = {}
+
+    def _get_python_function(self, function_id: int) -> Callable:
+        function, _ = self._python_functions.get(function_id, (None, None))
+        if function is None:
+            raise ValueError(
+                f"No Python function registered with id: {function_id}."
+            )
+        return function
+
+    def _get_python_function_file_name(self, function_id: int) -> str:
+        _, file_name = self._python_functions.get(function_id, (None, None))
+        if file_name is None:
+            raise ValueError(
+                f"No file found for Python function registered with id: {function_id}."
+            )
+        return file_name
```

Comment:
This single dictionary feels like it's doing a lot of different things, which leads to each of those things being a little clunky. More specifically, I don't think that we ever need both of the pieces inside of the dictionary value here at the same time, so putting them into a tuple together is a bit confusing. I think the two tasks that we're trying to accomplish here are:

- Keep track of a set of function ids that have been accounted for so far
- Once we're ready to serialize all the functions, do so and keep a map of id -> filepath.

We *could* merge these into one Map[int, Optional[String]], but I personally would prefer to have two separate fields since they don't really have much to do with one another and once we write the input files `None` really shouldn't be a valid value in the map. This should make these register and get functions much simpler.

Reply:
Refactored into two fields on the class!